OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs

```diff
@@ -42,23 +45,35 @@ public class WithAttackAnimation : ITick, INotifyAttack
 		readonly WithAttackAnimationInfo info;
 		readonly AttackBase attack;
 		readonly Armament armament;
-		readonly WithSpriteBody wsb;
+		readonly WithSpriteBody[] wsbs;
+		readonly bool onlyAttackAnim;
 
 		int tick;
+		bool attackAnimPlaying;
 
 		public WithAttackAnimation(ActorInitializer init, WithAttackAnimationInfo info)
 		{
 			this.info = info;
 			attack = init.Self.Trait<AttackBase>();
 			armament = init.Self.TraitsImplementing<Armament>()
 				.Single(a => a.Info.Name == info.Armament);
-			wsb = init.Self.Trait<WithSpriteBody>();
+			wsbs = init.Self.TraitsImplementing<WithSpriteBody>().Where(w => info.BodyNames.Contains(w.Info.Name)).ToArray();
+			onlyAttackAnim = string.IsNullOrEmpty(info.AimSequence) && string.IsNullOrEmpty(info.ReloadPrefix);
 		}
 
 		void PlayAttackAnimation(Actor self)
 		{
 			if (!string.IsNullOrEmpty(info.AttackSequence))
-				wsb.PlayCustomAnimation(self, info.AttackSequence, () => wsb.CancelCustomAnimation(self));
+			{
+				var enabledBodies = wsbs.Where(Exts.IsTraitEnabled);
+				if (!enabledBodies.Any())
+					return;
+
+				attackAnimPlaying = true;
+				foreach (var wsb in enabledBodies)
+					wsb.PlayCustomAnimation(self, info.AttackSequence,
+						() => { wsb.CancelCustomAnimation(self); attackAnimPlaying = false; });
```

Comment:
The behaviour here is subtle: right now if any animation is completed then it will restart *all* the animations, which IMO is more consistent than requiring *all* the animations to complete to restart *any*.

Option #3 is to maintain a `HashSet<WithSpriteBody>` or `bool[]` so we know exactly which animations to restart.  I'm not sure whether this is better ingame behaviour than @reaperrr's current implementation, however.